apps/e2e-tests/src/process-compose/e2e.spec.ts

```diff
@@ -1,166 +1,192 @@
+import {
+  LogChecker,
+  LogCheckerLive,
+  ProcessCompose,
+  ProcessComposeLive,
+  Sequencer,
+  SequencerLive,
+} from './types';
+
+describe.sequential('E2E Tests with process-compose', () => {
   const sequencerConfigUrl = 'http://127.0.0.1:5553/get_sequencer_config';
   const sequencerFeedsConfigUrl = 'http://127.0.0.1:5553/get_feeds_config';
   const network = 'ink_sepolia';
 
-  let sequencerConfig: SequencerConfigV2;
   let feedsConfig: NewFeedsConfig;
-  let ADFSConsumer;
   let feedIds: Array<bigint>;
+  let processCompose;
+  let ADFSConsumer;
   let initialPrices: Record<string, number>;
 
-  beforeAll(async () => {
-    await startEnvironment('example-setup-03');
-  });
-
-  afterAll(async () => {
-    await stopEnvironment();
+  beforeAll(() =>
+    pipe(
+      ProcessCompose,
+      Effect.provide(ProcessComposeLive),
+      Effect.tap(pc => (processCompose = pc)),
+      Effect.tap(pc => pc.start('example-setup-03')),
+      Effect.runPromise,
+    ),
+  );
+
+  afterAll(() => {
+    return pipe(processCompose.stop(), Effect.runPromise);
   });
 
-  test('Test processes state shortly after start', async () => {
-    const equal = await loopWhile(
-      (equal: boolean) => !equal,
-      async () => {
-        try {
-          const processes = await parseProcessesStatus();
-          deepStrictEqual(processes, expectedPCStatuses03);
-          return true;
-        } catch {
-          return false;
-        }
-      },
-      1000,
-      10,
-    );
-    expect(equal).toBe(true);
-  });
-
-  test('Test sequencer config is available and in correct format', async () => {
-    sequencerConfig = await fetchAndDecodeJSON(
-      SequencerConfigV2Schema,
-      sequencerConfigUrl,
-    );
-
-    feedsConfig = await fetchAndDecodeJSON(
-      NewFeedsConfigSchema,
-      sequencerFeedsConfigUrl,
-    );
-
-    expect(sequencerConfig).toBeTypeOf('object');
-
-    // Collect initial information for the feeds and their prices
-    const url = sequencerConfig.providers[network].url;
-    const contractAddress = sequencerConfig.providers[network]
-      .contract_address as `0x${string}`;
-    const allow_feeds = sequencerConfig.providers[network].allow_feeds;
-
-    feedIds = allow_feeds?.length
-      ? (allow_feeds as Array<bigint>)
-      : feedsConfig.feeds.map(feed => feed.id);
-
-    ADFSConsumer = AggregatedDataFeedStoreConsumer.createConsumerByRpcUrl(
-      contractAddress,
-      url,
-    );
-
-    initialPrices = feedIds.reduce(
-      (acc, feedId) => {
-        acc[feedId.toString()] = 0;
-        return acc;
-      },
-      {} as Record<string, number>,
-    );
-    initialPrices = await mapValuePromises(
-      initialPrices,
-      async (feedId, _) =>
-        await ADFSConsumer.getLatestSingleData(feedId).then(res =>
-          Number(res.slice(0, 50)),
+  it.effect('Test processes state shortly after start', () =>
+    Effect.gen(function* () {
+      const equal = yield* Effect.promise(() =>
+        loopWhile(
```

Comment:
This is a suggestion for future improvement.

You can try replacing `loopWhile` with [Effect.retry](https://effect.website/docs/error-management/retrying/#retry)

Example how we use it here: https://github.com/blocksense-network/blocksense/blob/25be02e1429c15d0c94b03a90b272ae194588a53/apps/social-verification/src/endpoints.ts#L165-L170